### DISCUSSION TITLE
How to use CompiledQuery with Async calls

### DISCUSSION THREAD

**Discussion by alexarchen** [linq2db/linq2db#3262] [Q&A]
CompiledQuery work fine with sync calls and increases speed significantly.
However when using it with async calls I get exceptions on some calls.
For example: 
`CompiledQuery.Compile((Database d, string q) =>  d.Positions.Where(p => p.Code== q).ToArrayAsync(default)); ` 
works well, 
whereas
`CompiledQuery.Compile((Database d, string q) =>  d.Positions.Where(p => p.Code == q).Set(p=>p.Value,0).UpdateAsync(default))` doesn't work, `await Invoke(...)` throws
LinqToDB.Linq.LinqException: 'Sequence ... cannot be converted to SQL.'

**Comment by MaceWindu** [comment]
Could you test it with recent version and create issue if it still fails?

**Reply by alexarchen** [reply]
Yes, it's still the same on 3.5.0.

**Comment by alexarchen** [comment]
As I understand It happens due to [CompiledQuery.cs#L93](https://github.com/linq2db/linq2db/blob/9097e0742e53726c0359eecd63a68b5512f6d9c8/Source/LinqToDB/CompiledQuery.cs#L93
)
```
if (expr.Method.DeclaringType == typeof(AsyncExtensions) &&
								expr.Method.GetCustomAttributes(typeof(AsyncExtensions.ElementAsyncAttribute), true).Length != 0)
```
This works well for FirstAsync, SumAsync,... But UpdateAsync, InsertAsync,... are defined in LinqExtensions so they aren't treated as async

**Comment by MaceWindu** [comment]
tracked by #3266
